File: sdk/dotnet/src/AlignTech.FeatureSDK/Evaluator.cs
using System.Text.Json;
using AlignTech.FeatureSDK.Models;

namespace AlignTech.FeatureSDK;

internal class Evaluator
{
    public EvaluationResult Evaluate(FlagConfig flag, AlignTechUser user)
    {
        var baseResult = new EvaluationResult
        {
            FlagKey = flag.FlagKey,
            Enabled = false,
            Reason = string.Empty
        };

        if (flag.Status != "active")
        {
            baseResult.Reason = "flag_inactive";
            return baseResult;
        }

        if (flag.Rollout == null || flag.Rollout.Count == 0)
        {
            baseResult.Reason = "no_rollout_config";
            return baseResult;
        }

        var rollout = flag.Rollout;
        var type = rollout.ContainsKey("type") ? rollout["type"]?.ToString() ?? "boolean" : "boolean";
        var enabled = false;

        switch (type)
        {
            case "whitelist":
                if (rollout.ContainsKey("userIds") && user.UserId != null)
                {
                    var userIds = GetStringList(rollout["userIds"]);
                    if (userIds.Count > 0)
                    {
                        enabled = MatchesAny(user.UserId, userIds);
                    }
                }
                break;

            case "boolean":
                if (rollout.ContainsKey("value"))
                {
                    enabled = rollout["value"] is JsonElement element
                        ? element.ValueKind == JsonValueKind.True
                        : Convert.ToBoolean(rollout["value"]);
                }
                break;

            default:
                enabled = false;
                break;
        }

        baseResult.Enabled = enabled;
        baseResult.Reason = $"rollout:{type}";
        return baseResult;
    }

    private static List<string> GetStringList(object? value)
    {
        if (value is JsonElement element && element.ValueKind == JsonValueKind.Array)
        {
            return element.EnumerateArray()
                .Where(e => e.ValueKind == JsonValueKind.String)
                .Select(e => e.GetString() ?? string.Empty)
                .Where(s => !string.IsNullOrEmpty(s))
                .ToList();
        }

        if (value is IEnumerable<object> enumerable)
        {
            return enumerable.Select(o => o.ToString() ?? string.Empty)
                .Where(s => !string.IsNullOrEmpty(s))
                .ToList();
        }

        return new List<string>();
    }

    private static bool MatchesAny(string value, List<string> patterns)
    {
        foreach (var pattern in patterns)
        {
            if (pattern == "*") return true;

            if (pattern.StartsWith("*") && pattern.EndsWith("*"))
            {
                var substr = pattern[1..^1];
                if (value.Contains(substr)) return true;
            }
            else if (pattern.StartsWith("*"))
            {
                if (value.EndsWith(pattern[1..])) return true;
            }
            else if (pattern.EndsWith("*"))
            {
                if (value.StartsWith(pattern[..^1])) return true;
            }
            else if (pattern == value)
            {
                return true;
            }
        }

        return false;
    }
}

File: sdk/javascript/src/Evaluator.ts
import { FlagConfig, AlignTechUser, EvaluationResult } from './types';

export class Evaluator {
  evaluate(flag: FlagConfig, user: AlignTechUser): EvaluationResult {
    const base = {
      flagKey: flag.flagKey,
      enabled: false,
      reason: '',
    };

    if (flag.status !== 'active') {
      return { ...base, reason: 'flag_inactive' };
    }

    const rollout = flag.rollout;
    if (!rollout || Object.keys(rollout).length === 0) {
      return { ...base, reason: 'no_rollout_config' };
    }

    const type = rollout.type || 'boolean';
    let enabled = false;

    switch (type) {
      case 'whitelist': {
        const userIds = rollout.userIds as string[] | undefined;
        if (userIds && userIds.length > 0 && user.userId) {
          enabled = this.matchesAny(user.userId, userIds);
        }
        break;
      }
      case 'boolean': {
        enabled = rollout.value === true;
        break;
      }
      default:
        enabled = false;
    }

    return { ...base, enabled, reason: `rollout:${type}` };
  }

  private matchesAny(value: string, patterns: string[]): boolean {
    for (const pattern of patterns) {
      if (pattern === '*') return true;
      if (pattern.startsWith('*') && pattern.endsWith('*')) {
        const substr = pattern.substring(1, pattern.length - 1);
        if (value.includes(substr)) return true;
      } else if (pattern.startsWith('*')) {
        if (value.endsWith(pattern.substring(1))) return true;
      } else if (pattern.endsWith('*')) {
        if (value.startsWith(pattern.substring(0, pattern.length - 1))) return true;
      } else if (pattern === value) {
        return true;
      }
    }
    return false;
  }
}
